### DISCUSSION TITLE
How to set configuration parameters

### DISCUSSION THREAD

**Discussion by windsOne** [quartznet/quartznet#650] [General] [answered]
How to set configuration parameters

Such as the number of threads  or LockOnInsert  paramters ?
Here is the help class I wrote.
My version 2.6

        public static void addJob<T>(string jobName, string jobGroupName, string triggerName, string triggerGroupName, string cron, JobDataMap datas) where T : IJob
        {
            try
            {
                IScheduler sched = StdSchedulerFactory.GetDefaultScheduler();
                JobBuilder jobBuilder = JobBuilder.Create<T>().WithIdentity(jobName, jobGroupName);
                if (datas != null)
                {
                    jobBuilder.SetJobData(datas);
                }
                IJobDetail jobDetail = jobBuilder.Build();
                TriggerBuilder triggerBuilder = TriggerBuilder.Create();
                triggerBuilder.WithIdentity(triggerName, triggerGroupName);
                triggerBuilder.StartNow();
                triggerBuilder.WithSchedule(CronScheduleBuilder.CronSchedule(cron));
                ITrigger trigger = triggerBuilder.Build();
                sched.ScheduleJob(jobDetail, trigger);
                if (!sched.IsShutdown)
                {
                    sched.Start();
                }
            }
            catch (Exception)
            {
            }
        }

        public static void addJob<T>(string jobName, string jobGroupName, string triggerName, string triggerGroupName, TimeSpan interval, JobDataMap datas) where T : IJob
        {
            try
            {
                IScheduler sched = StdSchedulerFactory.GetDefaultScheduler();
                JobBuilder jobBuilder = JobBuilder.Create<T>().WithIdentity(jobName, jobGroupName);
                if (datas != null && ((DirtyFlagMap<string, object>)datas).Count > 0)
                {
                    jobBuilder.SetJobData(datas);
                }
                IJobDetail jobDetail = jobBuilder.Build();
                TriggerBuilder triggerBuilder = TriggerBuilder.Create();
                ITrigger trigger = triggerBuilder.WithIdentity(triggerName, triggerGroupName).StartNow().WithSimpleSchedule(delegate (SimpleScheduleBuilder x)
                {
                    x.WithInterval(interval).RepeatForever();
                })
                    .Build();
                sched.ScheduleJob(jobDetail, trigger);
                if (!sched.IsShutdown)
                {
                    sched.Start();
                }
            }
            catch (Exception)
            {
            }
        }
        /// <summary>
        /// 修改Job的时间
        /// </summary>
        /// <param name="jobName"></param>
        /// <param name="jobGroupName"></param>
        /// <param name="triggerName"></param>
        /// <param name="triggerGroupName"></param>
        /// <param name="cron"></param>
		public static void modifyJobTime(string jobName, string jobGroupName, string triggerName, string triggerGroupName, string cron)
        {
            try
            {
                IScheduler sched = StdSchedulerFactory.GetDefaultScheduler();
                TriggerKey triggerKey = new TriggerKey(triggerName, triggerGroupName);
                ICronTrigger trigger2 = (ICronTrigger)sched.GetTrigger(triggerKey);
                if (trigger2 != null)
                {
                    string oldTime = trigger2.CronExpressionString;
                    if (!oldTime.Equals(cron))
                    {
                        TriggerBuilder triggerBuilder = TriggerBuilder.Create();
                        triggerBuilder.WithIdentity(triggerName, triggerGroupName);
                        triggerBuilder.StartNow();
                        triggerBuilder.WithSchedule(CronScheduleBuilder.CronSchedule(cron));
                        trigger2 = (ICronTrigger)triggerBuilder.Build();
                        sched.RescheduleJob(triggerKey, trigger2);
                    }
                }
            }
            catch (Exception)
            {
            }
        }

        public static void removeJob(string jobName, string jobGroupName, string triggerName, string triggerGroupName)
        {
            try
            {
                IScheduler sched = StdSchedulerFactory.GetDefaultScheduler();
                TriggerKey triggerKey = new TriggerKey(triggerName, triggerGroupName);
                sched.PauseTrigger(triggerKey);
                sched.UnscheduleJob(triggerKey);
                sched.DeleteJob(new JobKey(jobName, jobGroupName));
            }
            catch (Exception)
            {
            }
        }

        public static void AddJobListener(IJobListener jobListener, string jobName, string jobGroupName)
        {
            IScheduler sched = StdSchedulerFactory.GetDefaultScheduler();
            sched.ListenerManager.AddJobListener(jobListener, KeyMatcher<JobKey>.KeyEquals(new JobKey(jobName, jobGroupName)));
        }

        public static void AddJobListener(IJobListener jobListener, string jobGroupName)
        {
            IScheduler sched = StdSchedulerFactory.GetDefaultScheduler();
            sched.ListenerManager.AddJobListener(jobListener, GroupMatcher<JobKey>.GroupEquals(jobGroupName));
        }

        public static void AddJobListener(IJobListener jobListener, params IMatcher<JobKey>[] matchers)
        {
            IScheduler sched = StdSchedulerFactory.GetDefaultScheduler();
            sched.ListenerManager.AddJobListener(jobListener, matchers);
        }

        public static void RemoveJobListener(string linstenerName)
        {
            IScheduler sched = StdSchedulerFactory.GetDefaultScheduler();
            sched.ListenerManager.RemoveJobListener(linstenerName);
        }

        public static void AddAllJobListener(IJobListener jobListener)
        {
            IScheduler sched = StdSchedulerFactory.GetDefaultScheduler();
            sched.ListenerManager.AddJobListener(jobListener, GroupMatcher<JobKey>.AnyGroup());
        }

        public static void startJobs()
        {
            try
            {
                IScheduler sched = StdSchedulerFactory.GetDefaultScheduler();
                sched.Start();
            }
            catch (Exception)
            {
            }
        }

        public static void shutdownJobs()
        {
            try
            {
                IScheduler sched = StdSchedulerFactory.GetDefaultScheduler();
                if (!sched.IsShutdown)
                {
                    sched.Shutdown();
                }
            }
            catch (Exception)
            {
            }
        }

I need an operation that inserts a database with a large number of jobs at the same time, and needs to guarantee tens of thousands of operations.

What kind of changes do you need to write about this help class? Thank you, I can pay for it.

**Comment by lahma** [comment, marked as answer]
You can define quartz.config file that contains configuration to use, or define them inline for App.config. [See the server project as an example](https://github.com/quartznet/quartznet/tree/67697a1f610911d494808487f3b2af0f5b3724e7/src/Quartz.Server).

You should also keep the scheduler after you initialize it, so it should be sufficient for your use case to have global scheduler in static field or something and use that to operate.

**Comment by windsOne** [comment]
@lahma  
i change this type  ，but   ["quartz.jobStore.lockOnInsert"] = "false",   throw  error

        private QuartzManager()
        {

        }
        public static QuartzManager Instantation()
        {
            if (quartzManager == null)
            {
                quartzManager = new QuartzManager();
                sched = new StdSchedulerFactory(Prop()).GetScheduler();
            }
            return quartzManager;
        }


        private static NameValueCollection Prop()
        {
            NameValueCollection props = new NameValueCollection();
            props["lockOnInsert "] = "false";
            props["quartz.threadPool.threadCount"] = "20";
            return props;
        }

        private static IScheduler sched;
        private static QuartzManager quartzManager;
        public void addJob<T>(string jobName, string jobGroupName, string triggerName, string triggerGroupName, string cron, JobDataMap datas) where T : IJob
        {
            try
            {
                //IScheduler sched = new StdSchedulerFactory()
                //IScheduler sched = StdSchedulerFactory.GetDefaultScheduler();
                JobBuilder jobBuilder = JobBuilder.Create<T>().WithIdentity(jobName, jobGroupName);
                if (datas != null)
                {
                    jobBuilder.SetJobData(datas);
                }
                IJobDetail jobDetail = jobBuilder.Build();
                TriggerBuilder triggerBuilder = TriggerBuilder.Create();
                triggerBuilder.WithIdentity(triggerName, triggerGroupName);
                triggerBuilder.StartNow();
                triggerBuilder.WithSchedule(CronScheduleBuilder.CronSchedule(cron));
                ITrigger trigger = triggerBuilder.Build();
                sched.ScheduleJob(jobDetail, trigger);
                if (!sched.IsShutdown)
                {
                    sched.Start();
                }
            }
            catch (Exception)
            {
            }
        }

**Comment by lahma** [comment]
What is the error?

**Comment by windsOne** [comment]
@lahma 


2018/7/27 9:08:11 [INFO]  Quartz.Impl.StdSchedulerFactory - Using default implementation for object serializer
Quartz.SchedulerException: JobStore type '' props could not be configured. ---> Quartz.SchedulerConfigException: Could not parse property 'lockOnInsert' into correct data type: No writable property 'LockOnInsert' found ---> System.MemberAccessException: No writable property 'LockOnInsert' found
   在 Quartz.Util.ObjectUtils.SetPropertyValue(Object target, String propertyName, Object value) 位置 E:\Desktop\开源项 目\quartznet\src\Quartz\Util\ObjectUtils.cs:行号 185
   在 Quartz.Util.ObjectUtils.SetObjectProperties(Object obj, NameValueCollection props) 位置 E:\Desktop\开源项目\quartznet\src\Quartz\Util\ObjectUtils.cs:行号 150
   --- 内部异常堆栈跟踪的结尾 ---
   在 Quartz.Util.ObjectUtils.SetObjectProperties(Object obj, NameValueCollection props) 位置 E:\Desktop\开源项目\quartznet\src\Quartz\Util\ObjectUtils.cs:行号 154
   在 Quartz.Impl.StdSchedulerFactory.Instantiate() 位置 E:\Desktop\开源项目\quartznet\src\Quartz\Impl\StdSchedulerFactory.cs:行号 667
   --- 内部异常堆栈跟踪的结尾 ---
   在 Quartz.Impl.StdSchedulerFactory.Instantiate() 位置 E:\Desktop\开源项目\quartznet\src\Quartz\Impl\StdSchedulerFactory.cs:行号 672
   在 Quartz.Impl.StdSchedulerFactory.GetScheduler() 位置 E:\Desktop\开源项目\quartznet\src\Quartz\Impl\StdSchedulerFactory.cs:行号 1121
   在 ConsoleApplication3.QuartzManager.Instantation() 位置 E:\Desktop\work\项目\莱斯\源码\QuartzTry\QuartzTry\QuartzManager.cs:行号 21
   在 ConsoleApplication3.Program.Main(String[] args) 位置 E:\Desktop\work\项目\莱斯\源码\QuartzTry\QuartzTry\Program.cs:行号 22 [See nested exception: Quartz.SchedulerConfigException: Could not parse property 'lockOnInsert' into correct data type: No writable property 'LockOnInsert' found ---> System.MemberAccessException: No writable property 'LockOnInsert' found
   在 Quartz.Util.ObjectUtils.SetPropertyValue(Object target, String propertyName, Object value) 位置 E:\Desktop\开源项 目\quartznet\src\Quartz\Util\ObjectUtils.cs:行号 185
   在 Quartz.Util.ObjectUtils.SetObjectProperties(Object obj, NameValueCollection props) 位置 E:\Desktop\开源项目\quartznet\src\Quartz\Util\ObjectUtils.cs:行号 150
   --- 内部异常堆栈跟踪的结尾 ---
   在 Quartz.Util.ObjectUtils.SetObjectProperties(Object obj, NameValueCollection props) 位置 E:\Desktop\开源项目\quartznet\src\Quartz\Util\ObjectUtils.cs:行号 154
   在 Quartz.Impl.StdSchedulerFactory.Instantiate() 位置 E:\Desktop\开源项目\quartznet\src\Quartz\Impl\StdSchedulerFactory.cs:行号 667 [See nested exception: System.MemberAccessException: No writable property 'LockOnInsert' found
   在 Quartz.Util.ObjectUtils.SetPropertyValue(Object target, String propertyName, Object value) 位置 E:\Desktop\开源项 目\quartznet\src\Quartz\Util\ObjectUtils.cs:行号 185
   在 Quartz.Util.ObjectUtils.SetObjectProperties(Object obj, NameValueCollection props) 位置 E:\Desktop\开源项目\quartznet\src\Quartz\Util\ObjectUtils.cs:行号 150]]
Press any key to close the application





        private QuartzManager()
        {

        }
        public static QuartzManager Instantation()
        {
            if (quartzManager == null)
            {
                quartzManager = new QuartzManager();
                sched = new StdSchedulerFactory(Prop()).GetScheduler();
            }
            return quartzManager;
        }


        private static NameValueCollection Prop()
        {
            NameValueCollection props = new NameValueCollection();
            props["quartz.jobStore.lockOnInsert"] = "false";
            props["quartz.threadPool.threadCount"] = "20";
            return props;
        }

        private static IScheduler sched;
        private static QuartzManager quartzManager;
        public void addJob<T>(string jobName, string jobGroupName, string triggerName, string triggerGroupName, string cron, JobDataMap datas) where T : IJob
        {
            try
            {
                //IScheduler sched = new StdSchedulerFactory()
                //IScheduler sched = StdSchedulerFactory.GetDefaultScheduler();
                JobBuilder jobBuilder = JobBuilder.Create<T>().WithIdentity(jobName, jobGroupName);
                if (datas != null)
                {
                    jobBuilder.SetJobData(datas);
                }
                IJobDetail jobDetail = jobBuilder.Build();
                TriggerBuilder triggerBuilder = TriggerBuilder.Create();
                triggerBuilder.WithIdentity(triggerName, triggerGroupName);
                triggerBuilder.StartNow();
                triggerBuilder.WithSchedule(CronScheduleBuilder.CronSchedule(cron));
                ITrigger trigger = triggerBuilder.Build();
                sched.ScheduleJob(jobDetail, trigger);
                if (!sched.IsShutdown)
                {
                    sched.Start();
                }
            }
            catch (Exception)
            {
            }
        }

**Comment by lahma** [comment]
You need to configure AdoJobStore to be used, please follow the docs and samples: https://www.quartz-scheduler.net/documentation/quartz-2.x/tutorial/job-stores.html

    quartz.jobStore.type = Quartz.Impl.AdoJobStore.JobStoreTX, Quartz
